source/common/upstream/cluster_manager_impl.cc

```diff
@@ -206,16 +207,69 @@ ClusterManagerImpl::ClusterManagerImpl(const envoy::api::v2::Bootstrap& bootstra
         bootstrap.cluster_manager().upstream_bind_config().source_address());
   }
 
+  // Cluster loading happens in two phases: first all the primary clusters
+  // are loaded, and then all the secondary clusters are loaded. As it
+  // currently stands all non-EDS clusters are primary and only EDS
+  // clusters are secondary.
+  // This two phase loading is done for the same reason that we have
+  // a two phase cluster initialization sequence: secondary clusters
+  // might depend on primary clusters, and thus we need to load
+  // secondary cluster dependencies first.
   for (const auto& cluster : bootstrap.static_resources().clusters()) {
-    loadCluster(cluster, false);
+    // First load all the primary clusters.
+    if (cluster.type() != envoy::api::v2::Cluster::EDS) {
```

Comment:
I'm a little confused by this change and don't understand why it is necessary. Is this because in v2 config each EDS cluster individually sets up a subscription on an API source using generic code, so the non-EDS clusters must be loaded first? If that's the case can you update the comment to be more specific?

Reply:
Yes, exactly. I'll clarify the comment.